tests/unit/test_jenkins.py

```diff
@@ -1108,3 +1108,37 @@ def test_remove_unlisted_plugins(  # pylint: disable=too-many-arguments
         )
     else:
         mock_client.delete_plugins.assert_not_called()
+
+
+def test_rotate_credentials_error(
+    monkeypatch: pytest.MonkeyPatch, container: ops.Container, raise_exception: typing.Callable
+):
+    """
+    arrange: given a monkeypatched _invalidate_sessions that raises JenkinsAPIException.
+    act: when rotate_credentials is called.
+    assert: JenkinsError is raised.
+    """
+    monkeypatch.setattr(
+        jenkins,
+        "_invalidate_sessions",
+        lambda *_args, **_kwargs: raise_exception(
+            jenkinsapi.custom_exceptions.JenkinsAPIException
+        ),
```

Comment:
Perhaps instead of a `lambda` just use a mock with a side effect of raising an exception?

Reply:
Also applies to the rest of the `setattr`s with `lmabda`s

Reply:
Pushed on a separate PR to ease the reviews https://github.com/canonical/jenkins-k8s-operator/pull/62, thank you for the review!

Reply:
Merged to this branch.